openraft/src/engine.rs

```diff
@@ -0,0 +1,259 @@
+use std::collections::BTreeSet;
+use std::ops::Range;
+use std::sync::Arc;
+
+use maplit::btreeset;
+
+use crate::entry::RaftEntry;
+use crate::error::InitializeError;
+use crate::error::NotAllowed;
+use crate::storage::InitialState;
+use crate::EffectiveMembership;
+use crate::LogId;
+use crate::LogIdOptionExt;
+use crate::Membership;
+use crate::MetricsChangeFlags;
+use crate::NodeId;
+use crate::Vote;
+
+/// Raft protocol algorithm.
+///
+/// It implement the complete raft algorithm except does not actually update any states.
+/// But instead, it output commands to let a `RaftRuntime` implementation execute them to actually update the states
+/// such as append-log or save-vote by execute .
+///
+/// This structure only contains necessary information to run raft algorithm,
+/// but none of the application specific data.
+/// TODO: make the fields private
+#[derive(Debug, Clone, Default)]
+pub(crate) struct Engine<NID: NodeId> {
+    /// TODO:
+    #[allow(dead_code)]
+    pub(crate) id: NID,
+
+    /// Cache of a cluster of only this node.
+    ///
+    /// It is used to check an early commit if there is only one node in a cluster.
+    pub(crate) single_node_cluster: BTreeSet<NID>,
+
+    /// The state of this raft node.
+    pub(crate) state: InitialState<NID>,
+
+    /// Tracks what kind of metrics changed
+    pub(crate) metrics_flags: MetricsChangeFlags,
+
+    /// Command queue that need to be executed by `RaftRuntime`.
+    pub(crate) commands: Vec<Command<NID>>,
+}
+
+/// Commands to send to `RaftRuntime` to execute, to update the application state.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) enum Command<NID: NodeId> {
+    // Append a `range` of entries in the input buffer.
+    AppendInputEntries {
```

Comment:
emm...i read a lot to understand how `MoveInputCursorBy` be used. 

here is my  guess: if there is some entry from client, they must be append into log, this is done by `AppendInputEntries` cmd; but is the state is `leader`, then the entry also need to be replicated by `ReplicateInputEntries`, so using `MoveInputCursorBy` to control the buffer.am i right?

How about add a field in `AppendInputEntries` like `need_to_replicate` to notify the core if or not need to be replicated? in this way:

1. not Leader state must check `need_to_replicate` is false.
2. there is no need of `ReplicateInputEntries` and `MoveInputCursorBy` to control the input buffer, just the `AppendInputEntries` is needed, it is more clearly.

Reply:
> here is my guess: if there is some entry from client, they must be append into log, this is done by `AppendInputEntries` cmd; but is the state is `leader`, then the entry also need to be replicated by `ReplicateInputEntries`, so using `MoveInputCursorBy` to control the buffer.am i right?

Right. 

Another field like `need_to_replicate` is the best choice for the simplest cases:

Sometimes, to maximize the throughput, the input entries buffer may be a bit more complicated: There could be a Membership config entry in the mid in the `input_entries`.

E.g., assumes it is a single **member** cluster with several **learner**, and the input entries are `e1..e3,e4(membership{1,2,3}),e5,e6`:

`e1..e3` can be committed at once after being appended to log. And then to replicate them to learners.
`e4..e6` need to be replicated to a quorum in `{1,2,3}` and maybe to some of the learners then commit them.

The most efficient command sequence is:
```
append(e1..e6)
commit(e1..e3) // for quick commit
replicate(e1..e6)
move_cursor_to
```

Thus the `append` and `replicate` do not have to be two adjacent commands.

Reply:
in the example above, the pseudo-code may be these way:

```
let commit_entry_index = 0;
let replicate_entry = vec![];

for entry in input_entry {
  append_entry_to_log(entry);
  if entry = change_membership {
    change_membership(entry); // update_replication_streams
  }
  if entry_reach_commit {
    update commit_entry_index
 }
  if entry need to replicate {
    replicate_entry.append(entry);
  }
}

replicate_entry_to(replicate_entry);
```

Reply:
Correct. But the goal is to separate the raft algorithm and the runtime.
So that a deterministic test suite for the algorithm is possible.

This means the runtime does not involve any raft algorithm in it.